plutus-contract/src/Language/Plutus/Contract/Effects.hs

```diff
@@ -0,0 +1,100 @@
+import           Language.Plutus.Contract.Effects.ExposeEndpoint as X
+import           Language.Plutus.Contract.Effects.WatchAddress   as X
+import           Language.Plutus.Contract.Effects.WriteTx        as X
+
+import           Language.Plutus.Contract.Prompt.Event           (Event)
+import           Language.Plutus.Contract.Prompt.Hooks           (Hook)
+
+{- Note [Contract Effects]
+
+We model effects of Plutus contracts using 'Control.Eff'. The ones
+exported here are the basic effects that we need for our use cases (wait for a
+slot, expose an endpoint, watch an address on the blockchain, and submit
+transactions). To make these easier to use we export them as the
+'App' constraint.
+
+To our users, the signatures of our contracts look like '(Plutus r) =>
+Contract r ()'.
+
+Internally, all effects are interpreted as request-response pairs. The four
+effects 'AwaitSlot', 'ExposeEndpoint', 'WatchAddress' and 'WriteTx' have a
+common type for requests ('Hook') and a common type for responses ('Event').
+
+'runEffects' takes the user-facing contract request types and interprets them
+in '[Reader (Maybe Event), Exc (Hook ())]]' -- effectively 'Maybe Event -> Either (Hook ()) a'.
+
+-}
+
+-- | Basic interactions with the app platform (write transactions, wait for
+--   slot changes, watch addresses, and expose endpoints).
+type Plutus r = [WriteTx, AwaitSlot, WatchAddress, ExposeEndpoint] <:: r
+
+-- | List of effects that this interpreter ('Maybe Event -> Either Hooks a')
+--   supports.
+type PlutusEffects =
+    WriteTx ': AwaitSlot ': WatchAddress ': ExposeEndpoint ': PromptEffects
+
+type PromptEffects = '[Reader (Maybe Event), Exc (Hook ())]
+
+-- | Interpret the 'ContractEffects' in 'Reader' and 'Exc'. See note [Contract
+--   Effects]
+runEffects :: Eff PlutusEffects a -> Eff PromptEffects a
+runEffects =
+        runExposeEndpoint
+        . runWatchAddress
+        . runAwaitSlot
+        . runWriteTx
+
+{- note  [Hooks and Events]
+
+The three types 'Hook', 'Hooks' and 'Event' are closely related.
+
+* `Hook a` values are produced by the smart contract. They describe events that
+  the contract is waiting for. For example, `AddrHook :: Address -> Hook a`
+  means "I want to know the next transaction that affects this address". The
+  type parameter `a` is currently always the unit type, so it is effectively
+  unused. This parameter will hold information about the type (schema) of
+  endpoints that the contract exposes.
+
+* `Hooks` is a set of all hooks that the contract exposes in its current state.
+  The contract may have multiple branches running in parallel, and when a branch
+  gets blocked on a `Hook ()`, that hook is included in the `Hooks` value that
+  is returned to the client.
+
+* `Event` values are produced by the app platform. Each constructor of `Event`
+  corresponds to a constructor of `Hook a`. For example, `LedgerUpdate ::
+  Address -> Tx -> Event` is the counterpart of `AddrHook`.
```

Comment:
Yes ideally we should have an open union of request-response pairs, one for each of the effects of the contract. I'll leave that to https://github.com/input-output-hk/plutus/issues/1317 though (all our contracts have the same set of effect so, even though it would be nice to have now, we really only need it when we have other effects or different backends)